File: oxo2-shared/src/main/java/uk/ac/ebi/spot/oxo/model/sssom/WeakPredicate.java
package uk.ac.ebi.spot.oxo.model.sssom;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;

import java.util.List;

/**
 * A mapping predicate that is hidden from search results unless the user asks for it (ADR-0035).
 *
 * <p>Neither {@code rdfs:subClassOf} nor {@code oboInOwl:hasDbXref} asserts that two entities are the
 * same thing, and neither participates in SSSOM inference. On an OLS-derived corpus they are the
 * overwhelming majority of mappings, so showing them by default drowns the equivalences a user is
 * usually looking for. They are therefore excluded by default — but they are excluded, not absent,
 * and each can be switched back on independently: hierarchy and cross-references are different
 * questions and a user may want one without the other.
 *
 * <p>This enum is the single source of truth for the pair, shared by the two places that must agree
 * on them by construction:
 * <ul>
 *   <li>{@code SolrQueryBuilder}, which subtracts the enabled ones from the exclusion filter it puts
 *       on {@code oxo2-mappings}; and</li>
 *   <li>{@code Mappings2Entities}, which buckets every mapping sighting by predicate so the
 *       {@code oxo2-entities} typeahead can offer exactly the entities the search can then show.</li>
 * </ul>
 * If those two disagreed, the typeahead would suggest entities whose mappings the search hides — the
 * bug ADR-0035 exists to fix.
 *
 * <p>{@link #code()} is the wire form (API and URL); {@link #bucket()} names the entity-count fields
 * in {@code oxo2-entities} and so is lower-case, matching the Solr field-naming convention.
 */
public enum WeakPredicate {

    /** Ontology hierarchy, not equivalence. */
    SUB_CLASS_OF("subClassOf", "http://www.w3.org/2000/01/rdf-schema#subClassOf", "subclassof"),

    /** A loose cross-reference: no semantics beyond "these two are related somehow". */
    HAS_DB_XREF("hasDbXref", "http://www.geneontology.org/formats/oboInOwl#hasDbXref", "hasdbxref");

    private final String code;
    private final String iri;
    private final String bucket;

    WeakPredicate(String code, String iri, String bucket) {
        this.code = code;
        this.iri = iri;
        this.bucket = bucket;
    }

    /** The wire form, as it appears in the API and in the {@code wp} URL parameter. */
    @JsonValue
    public String code() {
        return code;
    }

    /**
     * The canonical predicate IRI. Filtering matches on {@code predicate_iri} rather than
     * {@code predicate_id}, whose CURIE prefix and casing vary by source set.
     */
    public String iri() {
        return iri;
    }

    /**
     * Names this predicate's count fields in {@code oxo2-entities}; see
     * {@code EntityConstants.subjectCountField}.
     */
    public String bucket() {
        return bucket;
    }

    /** The IRIs of the given predicates — what the mappings-side filter clauses are built from. */
    public static List<String> irisOf(List<WeakPredicate> predicates) {
        return predicates.stream().map(WeakPredicate::iri).toList();
    }

    /** Null rather than throwing, so a caller can turn an unknown value into a 400. */
    @JsonCreator
    public static WeakPredicate fromCode(String code) {
        if (code == null) {
            return null;
        }
        for (WeakPredicate predicate : values()) {
            if (predicate.code.equalsIgnoreCase(code.strip())) {
                return predicate;
            }
        }
        return null;
    }
}

File: oxo2-frontend/src/model/WeakPredicate.ts
// Weak predicates (ADR-0035): mapping predicates hidden from results unless the user asks for them.
// The codes match the backend WeakPredicate enum and travel as-is in the API and the `wp` URL param.
//
// Neither predicate says two entities are the same thing, and on an OLS-derived corpus they are the
// overwhelming majority of mappings — so showing them by default buries the equivalences people
// actually search for. They are hidden, not absent: each has its own checkbox, because ontology
// hierarchy and loose cross-references are different questions and wanting one is no reason to be
// shown the other.
//
// The same selection drives the typeahead. A suggestion is a promise that searching for that entity
// returns something, so the suggest is filtered by exactly these checkboxes — tick nothing and an
// entity whose every mapping is an xref is not offered at all, because picking it would land you on
// an empty table.

export type WeakPredicate = 'subClassOf' | 'hasDbXref';

export const WEAK_PREDICATE_ORDER: WeakPredicate[] = ['subClassOf', 'hasDbXref'];

export const WEAK_PREDICATE_LABELS: Record<WeakPredicate, string> = {
    subClassOf: 'Subclass (rdfs:subClassOf)',
    hasDbXref: 'Cross-reference (oboInOwl:hasDbXref)',
};

export const WEAK_PREDICATE_HINTS: Record<WeakPredicate, string> = {
    subClassOf: 'Ontology hierarchy, not equivalence.',
    hasDbXref: 'A loose link between two identifiers, with no stated meaning.',
};

// Hidden by default: both boxes start unchecked.
export const DEFAULT_WEAK_PREDICATES: WeakPredicate[] = [];

export function asWeakPredicate(value?: string | null): WeakPredicate | null {
    return value != null && value in WEAK_PREDICATE_LABELS ? (value as WeakPredicate) : null;
}
